app/src/main/kotlin/com/wire/android/ui/home/conversationslist/model/Conversation.kt

```diff
@@ -5,21 +5,21 @@ import com.wire.android.R
 import com.wire.android.ui.main.conversationlist.common.UserInfoLabel
 
 
-data class ConversationFolder(
-    val folderName: String,
-)
+abstract class Conversation {
+    abstract val conversationType: ConversationType
+}
 
-data class Conversation(
-    val userInfo: UserInfo,
-    val conversationInfo: ConversationInfo
-)
+sealed class ConversationType {
+    data class GroupConversation(
+        val groupColorValue: Long,
+        val groupName: String
+    ) : ConversationType()
 
-fun Conversation.toUserInfoLabel() =
-    UserInfoLabel(
-        labelName = conversationInfo.name,
-        isLegalHold = conversationInfo.isLegalHold,
-        membership = conversationInfo.membership,
-    )
+    data class PrivateConversation(
+        val userInfo: UserInfo,
+        val conversationInfo: ConversationInfo
+    ) : ConversationType()
+}
```

Comment:
What about just having a sealed class without open/abstract `Conversation`?

```kotlin
sealed class Conversation {
    data class GroupConversation(
        val groupColorValue: Long,
        val groupName: String
    ) : Conversation()

    data class PrivateConversation(
        val userInfo: UserInfo,
        val conversationInfo: ConversationInfo
    ) : Conversation()
}
```

This way we can get exhaustive `when` branches:
```kotlin
when(conversation){
    is GroupConversation -> {}
    is PrivateConversation -> {}
} //No need for an else branch
```

Reply:
that is actually what I did with :
```

sealed class ConversationType {
    data class GroupConversation(
        val groupColorValue: Long,
        val groupName: String
    ) : ConversationType()

fun Conversation.toUserInfoLabel() =
    UserInfoLabel(
        labelName = conversationInfo.name,
        isLegalHold = conversationInfo.isLegalHold,
        membership = conversationInfo.membership,
    )
    data class PrivateConversation(
        val userInfo: UserInfo,
        val conversationInfo: ConversationInfo
    ) : ConversationType()
}
```


but the problem is that I wanted to create a generic function that just "emits" the compose depending on what Row Item it is, so that we can just call one function, it could be either a "Call", "Mention" or "General" Row item 

you can see it here: 

```

@Composable
inline fun <reified Conversation> ConversationItemFactory(
    item: Conversation,
    eventType: EventType? = null,
    noinline onConversationItemClick: () -> Unit
) {
    when (item) {
        is GeneralConversation -> AllConversationItem(
            item,
            eventType,
            onConversationItemClick
        )
        is Call -> CallConversationItem(
            item,
            eventType,
            onConversationItemClick
        )
        is Mention -> MentionConversationItem(
            item,
            eventType,
            onConversationItemClick
        )
        else -> throw IllegalStateException("Illegal item type for ConversationItemFactory")
    }
}
```

Reply:
Kinda feels like the model is upside down.

Because `Mention` isn't a `Conversation`. `Call` isn't a `Conversation` either.

They can have a reference to a `Conversation` maybe?

For example:
```kotlin
sealed class Conversation {
    data class GroupConversation(
        val groupColorValue: Long,
        val groupName: String
    ) : Conversation()

    data class PrivateConversation(
        val userInfo: UserInfo,
        val conversationInfo: ConversationInfo
    ) : Conversation()
}

data class Call(val callInfo: CallInfo, val conversation: Conversation)
data class Mention(val mentionInfo: MentionInfo, val conversation: Conversation)
```

There's no need for `Mention` and `Call` to extend `Conversation`.

The way it currently is, it allows us to do weird stuff:
```kotlin
val groupConversation = GroupConversation(color, name)
val call = Call(callInfo, groupConversation)
//So far, so good.

val callConversation: Conversation = Mention(mentionInfo, call) 
// This doesn't make sense.

println(callConversation is GroupConversation) // returns false. Why? I passed a GroupConversation, I would expect this to be true.
```

Reply:
hmmm you got a point here I will give a tough abnout it tommorow, if you have an idea, please comment it here

Reply:
Even though the UI looks similar, the models aren't. There's no need to make them all extend the same thing. 

They are displayed in different lists, and represent different events / data structures. The only common thing they have is that they hold a reference to a Conversation.

But we don't have a need to display them all in the same list, and if we did, it doesn't mean that the model itself should change, but rather the view state - the model representing the view. 

Doesn't seem nice to mix core model entities so much - so early - in the project.